### DISCUSSION TITLE
zustand concurrent update issue

### DISCUSSION THREAD

**Discussion by zhaoyaoxin** [pmndrs/zustand#2515] [Bug report]
## zustand concurrent update issue
I'm having a problem when I'm modifying a state using zustand I call the update function twice in quick succession, but instead of initializing with the result of the first update the second update function uses the previous initial value causing the first update to be overwritten.
The original list is [1,2,3,4] The first update goes from [1,2,3,4] to [1,2,5,4] but the second update goes from [1,2,3,4] to [1,2,3,6] resulting in the final result of [1,2,3,6] The first update is overwritten.

![image](https://github.com/pmndrs/zustand/assets/91306716/5211f670-49df-4ad1-ba2b-87018275d3d1)

"react": "^18.2.0"
"react-dom": "^18.2.0"
 "zustand": "^4.3.8"
 "immer": "^10.0.2"

**Comment by dai-shi** [comment]
I guess you are facing a stale closure, but not sure. Can you create a reproduction?
- The reproduction should be very small,
- making the update function small, and
- if possible, without using immer.

**Reply by dai-shi** [reply]
Or maybe removing "async" fixes it.

**Reply by zhaoyaoxin** [reply]
Sorry, I can't provide a reproduction because the demo I wrote doesn't trigger the issue.
Here's my runtime log printout, you can see that the second start state is not the end state of the first.
![image](https://github.com/pmndrs/zustand/assets/91306716/1c2b6928-3677-49a4-9c6e-183d1f6d5a99)
![image](https://github.com/pmndrs/zustand/assets/91306716/5cb44a56-8cf3-4798-b306-e641a4386b7b)
![image](https://github.com/pmndrs/zustand/assets/91306716/803424d1-e241-448c-a53a-b321265a9b49)

**Reply by zhaoyaoxin** [reply]
Yes, when I deleted immer the problem stopped, but why?

**Reply by dbritto-dev** [reply]
@zhaoyaoxin would you mind trying with `replace: true`? I mean: `set(state => {}, true)` instead of `set(state => {}) `.
